### ISSUE TITLE
MKLDNN produce wrong gradient for conv in some edge cases

### ISSUE THREAD

**Issue by ailzhang** [pytorch/pytorch#25101]
To repro, build with `USE_MKLDNN=1`. 
```
import torch
import torch.nn.functional as F

input = torch.ones(1, 9, 32, 32, dtype=torch.float).requires_grad_()
weight = torch.ones(3, 9, 5, 5, dtype=torch.float).requires_grad_()
out = F.conv2d(input, weight, stride=3, padding=1)
out.sum().backward()

input_cuda = torch.ones(1, 9, 32, 32, dtype=torch.float, device='cuda').requires_grad_()
weight_cuda = torch.ones(3, 9, 5, 5, dtype=torch.float, device='cuda').requires_grad_()
out_cuda = F.conv2d(input_cuda, weight_cuda, stride=3, padding=1)
out_cuda.sum().backward()

print((input.grad != input_cuda.grad.cpu()).sum())
print(input.grad[0, 0, :, 28:31])
print(input_cuda.grad[0, 0, :, 28:31])
```
When built with `USE_MKLDNN=0`, we have the correct result(also matches cuda).

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh

**Comment by soumith** [comment]
cc: @Jianhui-Li @gujinghui

**Comment by Jianhui-Li** [comment]
@ailzhang  Thanks for the report. Could you please share on which platform (OS, CPU) you observed the difference and the results with and without MKL-DNN? 

@emfomenk

**Comment by emfomenk** [comment]
Thanks for report!
Seems like a bug on our side.
I submitted intel/mkl-dnn#542.

**Comment by ailzhang** [comment]
I'm on 
```
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.1 LTS
Release:        18.04
Codename:       bionic
```
with 
```
model name      : Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz
```

**Comment by XiaobingSuper** [comment]
@emankov , I test the case in PyTorch side using MKLDNN v0.20.4, it can solve this problem. Thanks!

**Comment by XiaobingSuper** [comment]
@ailzhang, yes, I test it on skx-6148, the new MKLDNN v0.20.4 can solve it, the log details is:
```
mkldnn_verbose,info,Intel MKL-DNN v0.20.4 (commit 043ebed3f288c29ae6ddcfb7f400ea441f3b5be9)
mkldnn_verbose,info,Detected ISA is Intel AVX-512 with AVX512BW, AVX512VL, and AVX512DQ extensions
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_oihw out:f32_Ohwi16o,num:1,3x9x5x5,0.059082
mkldnn_verbose,exec,convolution,jit:avx512_common,forward_training,fsrc:nchw fwei:Ohwi16o fbia:undef fdst:nChw16c,alg:convolution_direct,mb4_ic9oc3_ih14oh4kh5sh3dh0ph1_iw14ow4kw5sw3dw0pw1,0.0791016
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nChw16c out:f32_nchw,num:1,4x3x4x4,0.0539551
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nchw out:f32_nChw16c,num:1,4x3x4x4,0.0090332
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_oihw out:f32_OIhw16o16i,num:1,3x9x5x5,0.0620117
mkldnn_verbose,exec,convolution,jit:avx512_common,backward_data,fsrc:nChw16c fwei:OIhw16o16i fbia:undef fdst:nChw16c,alg:convolution_direct,mb4_ic9oc3_ih14oh4kh5sh3dh0ph1_iw14ow4kw5sw3dw0pw1,0.0251465
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nChw16c out:f32_nchw,num:1,4x9x14x14,0.059082
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nchw out:f32_nChw8c,num:1,4x9x14x14,0.013916
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_nchw out:f32_nChw8c,num:1,4x3x4x4,0.0100098
mkldnn_verbose,exec,convolution,jit:avx2,backward_weights,fsrc:nChw8c fwei:OIhw8i8o fbia:undef fdst:nChw8c,alg:convolution_direct,mb4_ic9oc3_ih14oh4kh5sh3dh0ph1_iw14ow4kw5sw3dw0pw1,0.175049
mkldnn_verbose,exec,reorder,simple:any,undef,in:f32_OIhw8i8o out:f32_oihw,num:1,3x9x5x5,0.0300293
tensor(0)

```

**Comment by XiaobingSuper** [comment]
@ailzhang , I will try it. Thanks!

**Comment by XiaobingSuper** [comment]
@ailzhang, I can reproduce the memory problem which your said even using the new release, it is our problem, we are now solve it and will upgrade MKLDNN soon. Thanks!

**Comment by ailzhang** [comment]
@XiaobingSuper FYI I added a bit more about transposed conv and also `dilation` for my local test case. Just want to let you know that the gist is updated in case you are interested. Thanks for looking into this!

**Comment by emankov** [comment]
> @emankov , I test the case in PyTorch side using MKLDNN v0.20.4, it can solve this problem. Thanks!

You meant @emfomenk, I believe.

**Comment by XiaobingSuper** [comment]
@ailzhang , we have upgraded MKLDNN to v0.20.5 in https://github.com/pytorch/pytorch/pull/25757, which can solve those reported bugs in this issue. I tested this patch and can passed all the cases. Thanks!

**Comment by ailzhang** [comment]
@XiaobingSuper Thank you so much for fixing it! We can close this issue via the upgrade.
